File: weird_captcha_gym/shared_runtime/app/mechanics/nonogram_denouement.js
(() => {
  "use strict";

  const MECHANIC_ID = "nonogram_denouement";
  const VALUE = {reset: 0, ink: 1, clear: -1};
  const SYMBOL = {NORTH: "↑", EAST: "→", SOUTH: "↓", WEST: "←"};
  let model = null;
  let cleanup = null;

  const esc = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;").replaceAll('"', "&quot;");
  const interaction = () => model.state.control_condition?.interaction || "full";

  function lineClues(line) {
    const result = [];
    let run = 0;
    [...line, 0].forEach((value) => {
      if (value === 1) run += 1;
      else if (run) { result.push(run); run = 0; }
    });
    return result;
  }

  function sameArray(left, right) {
    return left.length === right.length && left.every((value, index) => value === right[index]);
  }

  function plateStatus() {
    const size = model.puzzle.size;
    const decided = model.grid.flat().filter((value) => value !== 0).length;
    const complete = decided === size * size;
    const rowsValid = model.grid.every((row, index) => sameArray(lineClues(row), model.puzzle.row_clues[index]));
    const colsValid = Array.from({length: size}, (_, col) => (
      sameArray(lineClues(model.grid.map((row) => row[col])), model.puzzle.col_clues[col])
    )).every(Boolean);
    return {decided, complete, valid: complete && rowsValid && colsValid};
  }

  function clearFreshFailure() {
    if (!model.freshFailure) return;
    model.freshFailure = false;
    const verdict = document.querySelector(".nd-verdict");
    if (verdict) { verdict.className = "nd-verdict"; verdict.innerHTML = ""; }
    model.helpers.setReadout("", "idle");
  }

  function record(event) {
    clearFreshFailure();
    model.events.push({sequence: model.events.length + 1, ...event});
  }

  function cellClass(row, col, value, proof = false) {
    const classes = [proof ? "nd-proof-cell" : "nd-cell"];
    classes.push(value === 1 ? "is-ink" : value === -1 ? "is-clear" : "is-unknown");
    if ((col + 1) % 5 === 0 && col + 1 < model.puzzle.size) classes.push("is-major-col");
    if ((row + 1) % 5 === 0 && row + 1 < model.puzzle.size) classes.push("is-major-row");
    if (!proof && model.selected?.row === row && model.selected?.col === col) classes.push("is-selected");
    return classes.join(" ");
  }

  function clueText(clues) {
    return clues.length ? clues.map((value) => `<b>${value}</b>`).join("") : "<b>0</b>";
  }

  function plateMarkup() {
    const size = model.puzzle.size;
    const cells = model.grid.flatMap((line, row) => line.map((value, col) => `
      <button class="${cellClass(row, col, value)}" data-cell="${row}:${col}"
        aria-label="row ${row + 1}, column ${col + 1}, ${value === 1 ? "ink" : value === -1 ? "clear" : "undecided"}"
        ${model.developed ? "disabled" : ""}>
        <span></span>
      </button>`)).join("");
    return `<div class="nd-plate" style="--size:${size}">
      <div class="nd-plate-corner"><span>RUN</span><b>INDEX</b></div>
      <div class="nd-col-clues" style="--size:${size}">
        ${model.puzzle.col_clues.map((clues, index) => `<div data-col-clue="${index}">${clueText(clues)}</div>`).join("")}
      </div>
      <div class="nd-row-clues" style="--size:${size}">
        ${model.puzzle.row_clues.map((clues, index) => `<div data-row-clue="${index}">${clueText(clues)}</div>`).join("")}
      </div>
      <div class="nd-cells mode-${interaction()}" style="--size:${size}" data-grid>${cells}</div>
    </div>`;
  }

  function proxyMarkup() {
    if (interaction() !== "simplified" || model.developed) return "";
    const selected = model.selected ? `R${model.selected.row + 1} · C${model.selected.col + 1}` : "—";
    return `<div class="nd-mark-proxy" aria-label="selected cell mark controls">
      <span><b>${selected}</b></span>
      <button data-proxy-mark="ink" ${model.selected ? "" : "disabled"}><i></i>INK</button>
      <button data-proxy-mark="clear" ${model.selected ? "" : "disabled"}><i>×</i>CLEAR</button>
      <button data-proxy-mark="reset" ${model.selected ? "" : "disabled"}>RESET</button>
    </div>`;
  }

  function pressPanelMarkup() {
    return `<section class="nd-compose-panel">
      <div class="nd-section-label"><span>01</span><div><small>RUN MATRIX</small><h2>Plate</h2></div></div>
      <div class="nd-compose-body">
        ${plateMarkup()}
      </div>
      ${proxyMarkup()}
      <button class="nd-develop" data-develop ${model.developed ? "disabled" : ""}>
        <span>DEVELOP</span><i>→</i>
      </button>
    </section>`;
  }

  function markerFor(row, col) {
    return model.puzzle.markers.find((marker) => marker.row === row && marker.col === col) || null;
  }

  function proofGridMarkup() {
    const size = model.puzzle.size;
    const cells = model.grid.flatMap((line, row) => line.map((value, col) => {
      const marker = markerFor(row, col);
      return `<div class="${cellClass(row, col, value, true)}" data-proof-cell="${row}:${col}">
        ${marker ? `<span class="nd-ring" data-ring="${esc(marker.id)}"><i></i><b>${esc(marker.label.replace("RING ", ""))}</b></span>` : ""}
      </div>`;
    })).join("");
    return `<div class="nd-proof-frame">
      <div class="nd-proof-grid" data-proof-grid style="--size:${size}">${cells}<i class="nd-pulse" data-pulse></i></div>
    </div>`;
  }

  function answerMarkup() {
    if (!model.developed) return "";
    const options = model.puzzle.answer_options;
    if (interaction() === "simplified") {
      return `<div class="nd-answer-bank mode-simplified">
        ${options.map((direction) => `<button class="${model.answer === direction ? "is-selected" : ""}" data-answer-proxy="${direction}"><i>${SYMBOL[direction]}</i><span>${direction}</span></button>`).join("")}
      </div>`;
    }
    return `<div class="nd-answer-bank mode-full">
      <div class="nd-slug-rack">
        ${options.map((direction) => `<button class="nd-direction-slug ${model.answer === direction ? "is-selected" : ""}" data-direction-slug="${direction}"><i>${SYMBOL[direction]}</i><span>${direction}</span></button>`).join("")}
      </div>
      <div class="nd-answer-well" data-answer-well aria-label="direction answer well"><b>${model.answer ? `${SYMBOL[model.answer]} ${model.answer}` : "◇"}</b></div>
    </div>`;
  }

  function theatreMarkup() {
    return `<section class="nd-theatre-panel ${model.developed ? "is-developed" : "is-waiting"}">
      <div class="nd-section-label"><span>02</span><div><small>AFTERIMAGE</small><h2>Proof field</h2></div></div>
      ${model.developed ? proofGridMarkup() : `<div class="nd-waiting-well" aria-label="inactive proof field"><i></i></div>`}
      ${model.developed ? `<div class="nd-question"><h3>${esc(model.puzzle.question)}</h3></div>` : ""}
      ${answerMarkup()}
    </section>`;
  }

  function footerMarkup() {
    return `<footer class="nd-footer">
      <div class="nd-footer-seal" aria-hidden="true"><i></i><i></i><i></i></div>
      <div class="readout" data-status="idle"></div>
      <button data-certify ${model.answer && !model.terminal ? "" : "disabled"}>CERTIFY <i>↗</i></button>
    </footer>`;
  }

  function renderUi() {
    const start = model.loopStartedAt;
    model.helpers.app.innerHTML = `<section class="nonogram-denouement mode-${interaction()}" data-interaction="${interaction()}" data-challenge-id="${esc(model.state.challenge_id)}" data-fresh-failure="${model.freshFailure ? "true" : "false"}">
      <div class="nd-verdict"></div>
      <header class="nd-masthead">
        <div class="nd-mark"><i></i><span>THE<br>AFTERIMAGE<br>OFFICE</span></div>
        <div><h1>${esc(model.state.prompt)}</h1></div>
        <div class="nd-mode"><small>PLATE</small><b>${model.puzzle.size} × ${model.puzzle.size}</b></div>
      </header>
      <main>${pressPanelMarkup()}${theatreMarkup()}</main>
      ${footerMarkup()}
      ${model.helpers.cheatPanelTemplate()}
    </section>`;
    model.loopStartedAt = start;
    bindControls();
    model.helpers.installCheatPanel();
    if (model.freshFailure) showVerdict("fail");
    if (model.terminal) showVerdict("pass");
    ensureAnimation();
  }

  function applyCells(mode, coordinates, inputSource, pointerButton = null) {
    if (model.developed || !coordinates.length) return;
    const after = VALUE[mode];
    const cells = coordinates.map(({row, col}) => ({row, col, before: model.grid[row][col], after}));
    if (cells.every((cell) => cell.before === after)) return;
    cells.forEach((cell) => { model.grid[cell.row][cell.col] = after; });
    record({type: "mark", mode, cells, input_source: inputSource, ...(pointerButton ? {pointer_button: pointerButton} : {})});
    model.helpers.setReadout("", "idle");
    renderUi();
  }

  function segment(start, end) {
    if (!end) return [start];
    if (start.row === end.row) {
      const low = Math.min(start.col, end.col); const high = Math.max(start.col, end.col);
      return Array.from({length: high - low + 1}, (_, index) => ({row: start.row, col: low + index}));
    }
    if (start.col === end.col) {
      const low = Math.min(start.row, end.row); const high = Math.max(start.row, end.row);
      return Array.from({length: high - low + 1}, (_, index) => ({row: low + index, col: start.col}));
    }
    return null;
  }

  function parseCell(node) {
    if (!node?.dataset.cell) return null;
    const [row, col] = node.dataset.cell.split(":").map(Number);
    return {row, col};
  }

  function clearPreview() {
    document.querySelectorAll(".nd-cell.is-preview-ink,.nd-cell.is-preview-clear,.nd-cell.is-preview-reset").forEach((node) => {
      node.classList.remove("is-preview-ink", "is-preview-clear", "is-preview-reset");
    });
  }

  function paintPreview(cells, mode) {
    clearPreview();
    (cells || []).forEach(({row, col}) => document.querySelector(`[data-cell="${row}:${col}"]`)?.classList.add(`is-preview-${mode}`));
  }

  function bindFullGrid() {
    const grid = document.querySelector("[data-grid]");
    if (!grid || model.developed) return;
    let stroke = null;
    grid.addEventListener("contextmenu", (event) => event.preventDefault());
    grid.addEventListener("pointerdown", (event) => {
      const node = event.target.closest("[data-cell]");
      if (!node || ![0, 2].includes(event.button)) return;
      event.preventDefault();
      const mode = event.shiftKey ? "reset" : event.button === 2 ? "clear" : "ink";
      const pointerButton = event.shiftKey ? "shift" : event.button === 2 ? "right" : "left";
      const start = parseCell(node);
      stroke = {start, end: start, mode, pointerButton, pointerId: event.pointerId};
      grid.setPointerCapture?.(event.pointerId);
      paintPreview([start], mode);
    });
    grid.addEventListener("pointermove", (event) => {
      if (!stroke) return;
      const node = document.elementFromPoint(event.clientX, event.clientY)?.closest?.("[data-cell]");
      const end = parseCell(node);
      const cells = segment(stroke.start, end);
      if (cells) { stroke.end = end; paintPreview(cells, stroke.mode); }
    });
    const finish = () => {
      if (!stroke) return;
      const cells = segment(stroke.start, stroke.end) || [stroke.start];
      const {mode, pointerButton} = stroke;
      stroke = null;
      clearPreview();
      applyCells(mode, cells, "direct_grid_stroke", pointerButton);
    };
    grid.addEventListener("pointerup", finish);
    grid.addEventListener("pointercancel", () => { stroke = null; clearPreview(); });
  }

  function bindSimplifiedGrid() {
    if (model.developed) return;
    document.querySelectorAll("[data-cell]").forEach((button) => button.addEventListener("click", () => {
      model.selected = parseCell(button);
      clearFreshFailure();
      renderUi();
    }));
    document.querySelectorAll("[data-proxy-mark]").forEach((button) => button.addEventListener("click", () => {
      if (!model.selected) return;
      applyCells(button.dataset.proxyMark, [model.selected], "proxy_mark_button");
    }));
  }

  function startDeveloping() {
    if (model.developed) return;
    const status = plateStatus();
    if (!status.valid) return;
    model.developed = true;
    model.loopStartedAt = performance.now();
    record({type: "develop", input_source: "develop_button"});
    model.helpers.setReadout("", "pending");
    renderUi();
  }

  function chooseAnswer(direction, inputSource, gesture = null) {
    if (!model.developed || !model.puzzle.answer_options.includes(direction)) return;
    model.answer = direction;
    record({type: "answer", direction, input_source: inputSource, ...(gesture ? {gesture} : {})});
    model.helpers.setReadout("", "pending");
    renderUi();
  }

  function bindAnswerDrags() {
    const well = document.querySelector("[data-answer-well]");
    if (!well) return;
    document.querySelectorAll("[data-direction-slug]").forEach((slug) => {
      let drag = null;
      slug.addEventListener("pointerdown", (event) => {
        if (event.button !== 0) return;
        event.preventDefault();
        drag = {lastX: event.clientX, lastY: event.clientY, travel: 0, samples: 0, pointerId: event.pointerId};
        slug.setPointerCapture?.(event.pointerId);
        slug.classList.add("is-dragging");
      });
      slug.addEventListener("pointermove", (event) => {
        if (!drag) return;
        drag.travel += Math.hypot(event.clientX - drag.lastX, event.clientY - drag.lastY);
        drag.lastX = event.clientX; drag.lastY = event.clientY; drag.samples += 1;
        const box = slug.getBoundingClientRect();
        slug.style.transform = `translate(${event.clientX - (box.left + box.width / 2)}px, ${event.clientY - (box.top + box.height / 2)}px) rotate(-3deg)`;
      });
      slug.addEventListener("pointerup", (event) => {
        if (!drag) return;
        drag.travel += Math.hypot(event.clientX - drag.lastX, event.clientY - drag.lastY);
        drag.samples += 1;
        const box = well.getBoundingClientRect();
        const dropped = event.clientX >= box.left && event.clientX <= box.right && event.clientY >= box.top && event.clientY <= box.bottom;
        const proof = {start_direction: slug.dataset.directionSlug, travel_px: Number(drag.travel.toFixed(3)), sample_count: drag.samples, dropped_in_well: dropped};
        drag = null;
        slug.classList.remove("is-dragging"); slug.style.transform = "";
        if (dropped && proof.travel_px >= 40 && proof.sample_count >= 2) chooseAnswer(slug.dataset.directionSlug, "direction_slug_drag", proof);
      });
      slug.addEventListener("pointercancel", () => { drag = null; slug.classList.remove("is-dragging"); slug.style.transform = ""; });
    });
  }

  function bindControls() {
    if (interaction() === "full") bindFullGrid();
    else bindSimplifiedGrid();
    document.querySelector("[data-develop]")?.addEventListener("click", startDeveloping);
    document.querySelectorAll("[data-answer-proxy]").forEach((button) => button.addEventListener("click", () => chooseAnswer(button.dataset.answerProxy, "direction_proxy_button")));
    if (interaction() === "full") bindAnswerDrags();
    document.querySelector("[data-certify]")?.addEventListener("click", submit);
  }

  function animateProof() {
    model.animationFrame = null;
    if (!model?.developed || model.terminal) return;
    const pulse = document.querySelector("[data-pulse]");
    if (!pulse) return;
    const route = model.puzzle.route;
    const segmentMs = model.puzzle.pulse_segment_ms;
    const segmentCount = route.length - 1;
    const phase = ((performance.now() - model.loopStartedAt) % (segmentCount * segmentMs)) / segmentMs;
    const index = Math.min(segmentCount - 1, Math.floor(phase));
    const fraction = phase - index;
    const start = route[index]; const end = route[index + 1];
    const row = start.row + (end.row - start.row) * fraction;
    const col = start.col + (end.col - start.col) * fraction;
    pulse.style.left = `${(col + 0.5) / model.puzzle.size * 100}%`;
    pulse.style.top = `${(row + 0.5) / model.puzzle.size * 100}%`;
    pulse.dataset.segment = String(index);
    model.animationFrame = requestAnimationFrame(animateProof);
  }

  function ensureAnimation() {
    if (model.developed && !model.animationFrame && !model.terminal) model.animationFrame = requestAnimationFrame(animateProof);
  }

  function showVerdict(kind) {
    const node = document.querySelector(".nd-verdict");
    if (!node) return;
    node.className = `nd-verdict is-${kind}`;
    node.innerHTML = `<b>${kind.toUpperCase()}</b>`;
  }

  async function submit() {
    if (!model || model.submitting || model.terminal || !model.answer) return;
    const current = model;
    current.submitting = true;
    current.helpers.setReadout("", "pending");
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify({
          mechanic_id: current.state.mechanic_id,
          task_id: current.state.task_id,
          challenge_id: current.state.challenge_id,
          interaction_mode: interaction(),
          events: current.events,
          final_grid: current.grid,
          final_answer: current.answer,
          completed: true,
        }),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        current.terminal = true;
        if (current.animationFrame) cancelAnimationFrame(current.animationFrame);
        current.animationFrame = null;
        current.helpers.setReadout("PASS", "passed");
        showVerdict("pass");
      } else if (outcome.passed === false && outcome.state) {
        const helpers = current.helpers;
        await render(outcome.state, helpers, {freshFailure: true});
        model.helpers.setReadout("FAIL", "error");
        showVerdict("fail");
      } else {
        current.submitting = false;
        current.helpers.setReadout("FAIL", "error");
        showVerdict("fail");
      }
    } catch (_error) {
      if (model === current) {
        current.submitting = false;
        current.helpers.setReadout("", "idle");
      }
    }
  }

  async function render(state, helpers, options = {}) {
    cleanup?.();
    document.body.dataset.mechanic = "nonogram-denouement";
    model = {
      state,
      helpers,
      puzzle: JSON.parse(JSON.stringify(state.puzzle)),
      grid: Array.from({length: state.puzzle.size}, () => Array(state.puzzle.size).fill(0)),
      events: [],
      selected: null,
      developed: false,
      answer: null,
      loopStartedAt: 0,
      animationFrame: null,
      freshFailure: Boolean(options.freshFailure),
      terminal: false,
      submitting: false,
    };
    renderUi();
    cleanup = () => {
      if (model?.animationFrame) cancelAnimationFrame(model.animationFrame);
    };
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics[MECHANIC_ID] = {rootSelector: ".nonogram-denouement", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/nonogram_denouement.py
from __future__ import annotations

import copy
import math
from typing import Any, Iterable


MECHANIC_ID = "nonogram_denouement"
DIRECTIONS = {
    (-1, 0): "NORTH",
    (0, 1): "EAST",
    (1, 0): "SOUTH",
    (0, -1): "WEST",
}
CELL_VALUES = {"ink": 1, "clear": -1, "reset": 0}


def _fail(message: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": message}


def _bind(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> str | None:
    if any(str(item.get("mechanic_id") or "") != MECHANIC_ID for item in (payload, truth, public)):
        return "mechanic mismatch"
    for key in ("task_id", "challenge_id"):
        expected = str(truth.get(key) or "")
        if not expected or str(payload.get(key) or "") != expected or str(public.get(key) or "") != expected:
            return f"stale or mismatched {key}"
    return None


def _line_clues(line: Iterable[int]) -> list[int]:
    clues: list[int] = []
    run = 0
    for value in [*line, 0]:
        if value == 1:
            run += 1
        elif run:
            clues.append(run)
            run = 0
    return clues


def _direction(start: dict[str, Any], end: dict[str, Any]) -> str:
    delta = (int(end["row"]) - int(start["row"]), int(end["col"]) - int(start["col"]))
    if delta not in DIRECTIONS:
        raise ValueError("proof route contains a non-cardinal segment")
    return DIRECTIONS[delta]


def _contract(truth: dict[str, Any], public: dict[str, Any]) -> tuple[dict[str, Any], list[list[int]], str, str]:
    puzzle = truth.get("puzzle")
    if not isinstance(puzzle, dict) or public.get("puzzle") != puzzle:
        raise ValueError("public proof plate differs from replay contract")
    parameters = truth.get("parameters")
    if not isinstance(parameters, dict) or public.get("parameters") != parameters:
        raise ValueError("difficulty parameters differ from replay contract")
    condition = truth.get("control_condition")
    if condition != public.get("control_condition"):
        raise ValueError("public control condition differs from replay contract")
    if condition is not None and condition.get("difficulty_parameters") != parameters:
        raise ValueError("condition parameters differ from generated plate")
    interaction = str((condition or {}).get("interaction") or "full")
    if interaction not in {"simplified", "full"}:
        raise ValueError("interaction mode is invalid")

    size = puzzle.get("size")
    solution = truth.get("solution")
    if isinstance(size, bool) or not isinstance(size, int) or not 5 <= size <= 10:
        raise ValueError("plate size is invalid")
    if parameters.get("grid_size") != size:
        raise ValueError("grid_size does not match the plate")
    if (
        not isinstance(solution, list)
        or len(solution) != size
        or any(not isinstance(row, list) or len(row) != size for row in solution)
        or any(value not in {0, 1} for row in solution for value in row)
    ):
        raise ValueError("solution matrix is invalid")
    row_clues = [_line_clues(row) for row in solution]
    col_clues = [_line_clues(solution[row][col] for row in range(size)) for col in range(size)]
    if puzzle.get("row_clues") != row_clues or puzzle.get("col_clues") != col_clues:
        raise ValueError("visible clues do not describe the solution")

    route = puzzle.get("route")
    if not isinstance(route, list) or len(route) != parameters.get("route_steps"):
        raise ValueError("proof route length differs from the profile")
    for index, point in enumerate(route):
        if not isinstance(point, dict):
            raise ValueError("proof route point is invalid")
        row, col = point.get("row"), point.get("col")
        if any(isinstance(value, bool) or not isinstance(value, int) for value in (row, col)):
            raise ValueError("proof route coordinates are invalid")
        if not 0 <= row < size or not 0 <= col < size or solution[row][col] != 1:
            raise ValueError("proof route leaves the developed ink")
        if index:
            _direction(route[index - 1], point)
    question_index = truth.get("question_route_index")
    if isinstance(question_index, bool) or not isinstance(question_index, int) or not 1 <= question_index < len(route) - 1:
        raise ValueError("question route index is invalid")
    correct_direction = _direction(route[question_index], route[question_index + 1])
    if truth.get("correct_direction") != correct_direction:
        raise ValueError("answer direction differs from the proof route")
    markers = puzzle.get("markers")
    if not isinstance(markers, list) or len(markers) != parameters.get("marker_count"):
        raise ValueError("marker count differs from the profile")
    target_marker_id = str(puzzle.get("target_marker_id") or "")
    target = next((item for item in markers if isinstance(item, dict) and item.get("id") == target_marker_id), None)
    if target is None or target.get("row") != route[question_index]["row"] or target.get("col") != route[question_index]["col"]:
        raise ValueError("question marker is not on the audited route point")
    options = puzzle.get("answer_options")
    if (
        not isinstance(options, list)
        or len(options) != parameters.get("answer_direction_count")
        or len(set(options)) != len(options)
        or correct_direction not in options
        or any(option not in DIRECTIONS.values() for option in options)
    ):
        raise ValueError("answer options are invalid")
    return puzzle, copy.deepcopy(solution), interaction, correct_direction


def _integer(value: Any, label: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ValueError(f"{label} must be an integer")
    return value


def _straight_contiguous(cells: list[dict[str, Any]]) -> bool:
    points = [(cell["row"], cell["col"]) for cell in cells]
    if len(set(points)) != len(points):
        return False
    rows = {point[0] for point in points}
    cols = {point[1] for point in points}
    if len(rows) == 1:
        ordered = sorted(point[1] for point in points)
    elif len(cols) == 1:
        ordered = sorted(point[0] for point in points)
    else:
        return False
    return ordered == list(range(ordered[0], ordered[-1] + 1))


def _validate_answer_gesture(event: dict[str, Any], direction: str) -> None:
    gesture = event.get("gesture")
    if not isinstance(gesture, dict):
        raise ValueError("direct answer lacks drag proof")
    if gesture.get("start_direction") != direction or gesture.get("dropped_in_well") is not True:
        raise ValueError("direct answer drag endpoints are invalid")
    travel = gesture.get("travel_px")
    samples = gesture.get("sample_count")
    if (
        isinstance(travel, bool)
        or not isinstance(travel, (int, float))
        or not math.isfinite(float(travel))
        or float(travel) < 40
        or isinstance(samples, bool)
        or not isinstance(samples, int)
        or samples < 2
    ):
        raise ValueError("direct answer drag is too short or sparsely sampled")


def grade(payload: dict[str, Any], truth: dict[str, Any], public: dict[str, Any]) -> dict[str, Any]:
    binding = _bind(payload, truth, public)
    if binding:
        return _fail(binding)
    try:
        puzzle, solution, interaction, correct_direction = _contract(truth, public)
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"invalid nonogram contract: {exc}")
    if payload.get("interaction_mode") != interaction:
        return _fail("submitted interaction mode differs from task condition")
    events = payload.get("events")
    size = puzzle["size"]
    if not isinstance(events, list) or not 3 <= len(events) <= size * size * 5 + 20:
        return _fail("interaction transcript is missing or outside limits")

    board = [[0] * size for _ in range(size)]
    developed = False
    answer: str | None = None
    mark_sources = {"full": "direct_grid_stroke", "simplified": "proxy_mark_button"}
    answer_sources = {"full": "direction_slug_drag", "simplified": "direction_proxy_button"}
    mark_count = 0
    try:
        for sequence, event in enumerate(events, 1):
            if not isinstance(event, dict) or event.get("sequence") != sequence:
                raise ValueError(f"event {sequence} has an invalid sequence")
            event_type = event.get("type")
            if event_type == "mark":
                if developed:
                    raise ValueError(f"event {sequence} edits a developed plate")
                if event.get("input_source") != mark_sources[interaction]:
                    raise ValueError(f"event {sequence} uses the wrong mark input surface")
                mode = str(event.get("mode") or "")
                if mode not in CELL_VALUES:
                    raise ValueError(f"event {sequence} has an invalid mark mode")
                cells = event.get("cells")
                if not isinstance(cells, list) or not 1 <= len(cells) <= size:
                    raise ValueError(f"event {sequence} has an invalid mark span")
                if interaction == "simplified" and len(cells) != 1:
                    raise ValueError(f"event {sequence} batches proxy marks")
                if interaction == "full":
                    if not _straight_contiguous(cells):
                        raise ValueError(f"event {sequence} is not a straight contiguous stroke")
                    expected_button = {"ink": "left", "clear": "right", "reset": "shift"}[mode]
                    if event.get("pointer_button") != expected_button:
                        raise ValueError(f"event {sequence} disagrees with its pointer button")
                for cell in cells:
                    if not isinstance(cell, dict):
                        raise ValueError(f"event {sequence} contains an invalid cell")
                    row = _integer(cell.get("row"), "cell row")
                    col = _integer(cell.get("col"), "cell col")
                    before = _integer(cell.get("before"), "cell before")
                    after = _integer(cell.get("after"), "cell after")
                    if not 0 <= row < size or not 0 <= col < size:
                        raise ValueError(f"event {sequence} leaves the plate")
                    if before != board[row][col] or after != CELL_VALUES[mode]:
                        raise ValueError(f"event {sequence} starts from stale cell state")
                    board[row][col] = after
                mark_count += 1
            elif event_type == "develop":
                if developed or event.get("input_source") != "develop_button":
                    raise ValueError(f"event {sequence} has an invalid develop transition")
                if any(value == 0 for row in board for value in row):
                    raise ValueError(f"event {sequence} develops an undecided plate")
                ink = [[1 if value == 1 else 0 for value in row] for row in board]
                if ink != solution:
                    raise ValueError(f"event {sequence} develops a clue-inconsistent plate")
                developed = True
            elif event_type == "answer":
                if not developed or event.get("input_source") != answer_sources[interaction]:
                    raise ValueError(f"event {sequence} uses an invalid answer surface or phase")
                direction = str(event.get("direction") or "")
                if direction not in puzzle["answer_options"]:
                    raise ValueError(f"event {sequence} chooses an unavailable direction")
                if interaction == "full":
                    _validate_answer_gesture(event, direction)
                answer = direction
            else:
                raise ValueError(f"event {sequence} has an unknown type")
    except (KeyError, TypeError, ValueError) as exc:
        return _fail(f"nonogram replay rejected: {exc}")

    if payload.get("final_grid") != board:
        return _fail("submitted plate does not match transcript replay")
    if payload.get("final_answer") != answer:
        return _fail("submitted direction does not match transcript replay")
    complete = payload.get("completed") is True and developed and answer is not None
    passed = complete and answer == correct_direction
    return {
        "graded": True,
        "passed": passed,
        "feedback": f"plate {size}×{size} exact; {mark_count} replayed mark strokes; developed={developed}; direction {answer or 'NONE'}",
    }
